source/mock/ur_mock_helpers.hpp

```diff
@@ -0,0 +1,129 @@
+    size_t MSize;
+};
+
+using dummy_handle_t = dummy_handle_t_ *;
+
+// Allocates a dummy handle of type T with support of reference counting.
+// Takes optional 'Size' parameter which can be used to allocate additional
+// memory. The handle has to be deallocated using 'releaseDummyHandle'.
+template <class T> inline T createDummyHandle(size_t Size = 0) {
+    dummy_handle_t DummyHandlePtr = new dummy_handle_t_(Size);
+    return reinterpret_cast<T>(DummyHandlePtr);
+}
+
+// Allocates a dummy handle of type T with support of reference counting
+// and associates it with the provided Data.
+template <class T>
+inline T createDummyHandleWithData(unsigned char *Data, size_t Size) {
+    auto DummyHandlePtr = new dummy_handle_t_(Data, Size);
+    return reinterpret_cast<T>(DummyHandlePtr);
+}
+
+// Decrement reference counter for the handle and deallocates it if the
+// reference counter becomes zero
+template <class T> inline void releaseDummyHandle(T Handle) {
+    auto DummyHandlePtr = reinterpret_cast<dummy_handle_t>(Handle);
+    const size_t NewValue = --DummyHandlePtr->MRefCounter;
+    if (NewValue == 0) {
+        delete DummyHandlePtr;
+    }
+}
+
+// Increment reference counter for the handle
+template <class T> inline void retainDummyHandle(T Handle) {
+    auto DummyHandlePtr = reinterpret_cast<dummy_handle_t>(Handle);
+    ++DummyHandlePtr->MRefCounter;
+}
+
+struct callbacks_t {
+    void set_before_callback(std::string name, ur_mock_callback_t callback) {
+        beforeCallbacks[name] = callback;
+    }
+
+    ur_mock_callback_t get_before_callback(std::string name) {
+        auto callback = beforeCallbacks.find(name);
+
+        if (callback != beforeCallbacks.end()) {
+            return callback->second;
+        }
+        return nullptr;
+    }
+
+    void set_replace_callback(std::string name, ur_mock_callback_t callback) {
+        replaceCallbacks[name] = callback;
+    }
+
+    ur_mock_callback_t get_replace_callback(std::string name) {
+        auto callback = replaceCallbacks.find(name);
+
+        if (callback != replaceCallbacks.end()) {
+            return callback->second;
+        }
+        return nullptr;
+    }
+
+    void set_after_callback(std::string name, ur_mock_callback_t callback) {
+        afterCallbacks[name] = callback;
+    }
+
+    ur_mock_callback_t get_after_callback(std::string name) {
+        auto callback = afterCallbacks.find(name);
+
+        if (callback != afterCallbacks.end()) {
+            return callback->second;
+        }
+        return nullptr;
+    }
+
+    void resetCallbacks() {
+        beforeCallbacks.clear();
+        replaceCallbacks.clear();
+        afterCallbacks.clear();
+    }
+
+  private:
+    std::unordered_map<std::string, ur_mock_callback_t> beforeCallbacks;
+    std::unordered_map<std::string, ur_mock_callback_t> replaceCallbacks;
+    std::unordered_map<std::string, ur_mock_callback_t> afterCallbacks;
+};
+
+extern callbacks_t callbacks;
```

Comment:
nit: I think I'd prefer an exported function with implementation like:
```
getCallbacks() {
  static callbacks_t callbacks;
  return callbacks;
}
```

Reply:
I don't mind making the change but what's the advantage? Something to do with how it's linked?

Reply:
Well, personally I don't like variables as part of public interfaces, a function seems better. Using a static inside of a function also abstracts the global nature of this type little better. AFAIR, on Windows there are different rules on how public variables and public functions are exported (see [docs](https://learn.microsoft.com/en-us/cpp/build/reference/exports?view=msvc-170)).

And finally, what I suggested is inline with [C++ core guidelines](https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#i3-avoid-singletons).

Reply:
made the change